public/app/features/dashboard/panel_editor/QueryEditorRows.tsx

```diff
@@ -42,27 +42,26 @@ export class QueryEditorRows extends PureComponent<Props> {
     panel.refresh();
   };
 
-  onChangeQuery(query: DataQuery, index: number) {
+  onChangeQuery = (query: DataQuery) => {
     const { queries, onChangeQueries } = this.props;
 
-    const old = queries[index];
+    // Assuming datasources won't mutate queries, we find the index of the query that was updated.
+    const prevQueryIndex = queries.findIndex((_, i, oldQueries) => oldQueries[i] !== queries[i]);
```

Comment:
How is this supposed to work? `oldQueries` is `queries` the third arg is just the whole array you are iterating over. So this should always return -1 or am I missing something?

Reply:
we need the index, query editors SHOULD pass a new updated query model instance to onChange (Query editors should not modify the props query model instance)

Reply:
It was clearly too late last night, sorry.
I didn't push the correct version, forget the whole `mutate queries` story, the actual solution was comparing queries' `refId` assuming query editors don't mutate it.

I'll update the PR